osu.Game/Database/RealmContextFactory.cs

```diff
@@ -374,32 +397,41 @@ public IDisposable BlockAllOperations()
             if (isDisposed)
                 throw new ObjectDisposedException(nameof(RealmContextFactory));
 
-            if (!ThreadSafety.IsUpdateThread)
-                throw new InvalidOperationException(@$"{nameof(BlockAllOperations)} must be called from the update thread.");
-
-            Logger.Log(@"Blocking realm operations.", LoggingTarget.Database);
-
             try
             {
                 contextCreationLock.Wait();
 
                 lock (contextLock)
                 {
+                    if (!ThreadSafety.IsUpdateThread && context != null)
+                        throw new InvalidOperationException(@$"{nameof(BlockAllOperations)} must be called from the update thread.");
+
+                    Logger.Log(@"Blocking realm operations.", LoggingTarget.Database);
+
                     context?.Dispose();
                     context = null;
                 }
 
                 const int sleep_length = 200;
                 int timeout = 5000;
 
-                // see https://github.com/realm/realm-dotnet/discussions/2657
-                while (!Compact())
+                try
                 {
-                    Thread.Sleep(sleep_length);
-                    timeout -= sleep_length;
+                    // see https://github.com/realm/realm-dotnet/discussions/2657
+                    while (!Compact())
+                    {
+                        Thread.Sleep(sleep_length);
+                        timeout -= sleep_length;
 
-                    if (timeout < 0)
-                        throw new TimeoutException(@"Took too long to acquire lock");
+                        if (timeout < 0)
+                            throw new TimeoutException(@"Took too long to acquire lock");
+                    }
+                }
+                catch (Exception e)
```

Comment:
This is also catching the inner `TimeoutException`, which makes the blocking not work (because if the attempt to block fails on a timeout, the timeout will get caught, logged, and the operation will go on its merry way), and also [fails a test](https://github.com/ppy/osu/blob/b3983db69d5a5646cae2c3bdd9db326ea472fbfd/osu.Game.Tests/Database/GeneralUsageTests.cs#L91-L96).

I think the thing to do here is to explicitly handle and re-throw the timeout? i.e.:

```suggestion
                catch (TimeoutException)
                {
                    throw;
                }
                catch (Exception e)
```

Reply:
Or alternatively catch `RealmException`s specifically rather than `Exception`.

Reply:
I've gone with scoping the catch further. I think it reads better.